engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -380,6 +513,122 @@ private void handleCreateTemplateFromSnapshot(SnapshotInfo snapshotInfo, Templat
+        _volumeService.updateHypervisorSnapshotReserveForVolume(diskOffering, volumeInfo.getId(), snapshot.getHypervisorType());
+
+
+        CopyCmdAnswer copyCmdAnswer = null;
+        String errMsg = null;
+
+        HostVO hostVO = null;
+        try {
+
+            //create a volume on the storage
+            AsyncCallFuture<VolumeApiResult> future = _volumeService.createVolumeAsync(volumeInfo, volumeInfo.getDataStore());
+            VolumeApiResult result = future.get();
+
+            if (result.isFailed()) {
+                LOGGER.warn("Failed to create a volume: " + result.getResult());
+                throw new CloudRuntimeException(result.getResult());
+            }
+
+            volumeInfo = _volumeDataFactory.getVolume(volumeInfo.getId(), volumeInfo.getDataStore());
+
+            volumeInfo.processEvent(Event.MigrationRequested);
+
+            volumeInfo = _volumeDataFactory.getVolume(volumeInfo.getId(), volumeInfo.getDataStore());
+
+            hostVO = getHost(snapshotInfo.getDataCenterId(), false);
+
+            //copy the volume from secondary via the hypervisor
+            copyCmdAnswer = performCopyOfVdi(volumeInfo, snapshotInfo, hostVO);
+
+            if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
+                if (copyCmdAnswer != null && !StringUtils.isEmpty(copyCmdAnswer.getDetails())) {
+                    errMsg = copyCmdAnswer.getDetails();
+                }
+                else {
+                    errMsg = "Unable to create volume from snapshot";
+                }
+            }
+        }
+        catch (Exception ex) {
+            errMsg = ex.getMessage() != null ? ex.getMessage() : "Copy operation failed in 'StorageSystemDataMotionStrategy.handleCreateVolumeFromSnapshotBothOnStorageSystem'";
+        } finally {
+
+            // detach and remove access after the volume is created
+            DiskTO disk = new DiskTO(volumeInfo.getTO(), volumeInfo.getDeviceId(), volumeInfo.getPath(), volumeInfo.getVolumeType());
+            DettachCommand cmd = new DettachCommand(disk, null);
+            long storagePoolId = volumeInfo.getPoolId();
+            StoragePoolVO storagePool = _storagePoolDao.findById(storagePoolId);
+
+            cmd.setManaged(true);
+            cmd.setStorageHost(storagePool.getHostAddress());
+            cmd.setStoragePort(storagePool.getPort());
+            cmd.set_iScsiName(volumeInfo.get_iScsiName());
+
+            try {
+                DettachAnswer dettachAnswer = (DettachAnswer) _agentMgr.send(hostVO.getId(), cmd);
+
+                if (!dettachAnswer.getResult()) {
+                    throw new CloudRuntimeException("Error detaching volume:" + dettachAnswer.getDetails());
+                }
+
+            } catch (Exception e) {
+                LOGGER.warn("Error detaching volume " + volumeInfo.getId() + " Error: " + e.getMessage());
+            }
+
+            _volumeService.revokeAccess(volumeInfo, hostVO, destDataStore);
```

Comment:
Lines 571-593 appear to duplicate Line 463-475 in ``StorageSystemDataMotionStrategy``.  Please consider consolidation of the commonalities.